rusoto/core/src/pagination.rs

```diff
@@ -0,0 +1,76 @@
+//! Traits and helperfunctions for dealing with paged responses
+
+use futures::{stream, Stream, TryStreamExt, Future};
+use crate::{RusotoError};
+use std::pin::Pin;
+
+/// A stream of results returned over multiple pages by amazon
+pub type RusotoStream<I, E> = Pin<Box<dyn Stream<Item = Result<I, RusotoError<E>>>>>;
+
+/// Trait representing a request that can have a token added to fetch subsequent pages
+pub trait PagedRequest {
+    /// The type of the paging token, usuaully `Option<String>` but sometimes a tuple or something mor exotic
+    type Token;
+
+    /// Set the token to get the next page
+    /// See [PagedOutput::pagination_token]
+    fn with_pagination_token(self, key: Self::Token) -> Self;
+} 
+
+
+/// Trait representing the partial output of a request that may have more pages of results 
+pub trait PagedOutput {
```

Comment:
This could have a simpler API I think, say

```
fn split_into_paginated(self) -> (Option<Self::Token>, Vec<Self::Item>);
```

Reply:
I feel being able to ask for the token separately is useful, but to reduce the amount of cloning I added a "take" version that is similar to `Option::take`.